backupccl: elide data from offline tables during RESTORE

msbutler opened the pull request:
A previous PR (https://github.com/cockroachdb/cockroach/pull/83915) enabled BACKUP to target offline descriptors
(including IMPORTing and RESTOREing tables), but didn't fully handle their
rollback on RESTORE. This PR ensures RESTORE properly handles these offline
tables:
- An offline table created by RESTORE or IMPORT PGDUMP is fully discarded. The
  table does not exist in the restoring cluster.
- An offline table created by an IMPORT INTO has all importing data elided in
  the restore processor and is restored online to its pre import state.

If the RESTORE occurs AOST after the table returns online, the table is
RESTOREd to its post offline state.

Currently, all tables that return online are fully re-backed up. For many
situations above, this is unecessary. Future work should ensure that these
tables are not re-backed up when we know all operations within the
incremental backup interval on the table were MVCC.

This patch only applies to BACKUP and RESTORE flavors that implicitly target an
offline importing/restoring table. A future PR should allow a user to
explicitly target an existing offline table undergoing an IMPORT.

Fixes #86054

Release justification: fixes bug where offline tables were not properly handled
in restore.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/85920)
<!-- Reviewable:end -->

msbutler commented:
RFAL! all test failures relate to first commit which relates to #85979, and not this PR.

dt commented on pkg/ccl/backupccl/key_rewriter.go:
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
+
+	// Elide Importing Keys
+	if importTime := desc.GetInProgressImportStartTime(); wallTime > 0 && importTime > 0 && wallTime >= importTime {
+		return nil, false, ErrImportingKeyError
```
do you need a non-nil err? isn't false enough?

msbutler committed:
backupccl: elide data from offline tables in RESTORE

A previous PR (#83915) enabled BACKUP to target offline descriptors
(including IMPORTing and RESTOREing tables), but didn't fully handle their
rollback on RESTORE. This PR ensures RESTORE properly handles these offline
tables:
- An offline table created by RESTORE or IMPORT PGDUMP is fully discarded. The
  table does not exist in the restoring cluster.
- An offline table created by an IMPORT INTO has all importing data elided in
  the restore processor and is restored online to its pre import state.

If the RESTORE occurs AOST after the table returns online, the table is
RESTOREd to it's post offline state.

Currently, all tables that return online are fully re-backed up. For many
situations above, this is unecessary. Future work should ensure that these
tables are not re-backed up when we know all operations within the
incremental backup interval on the table were MVCC.

This patch only applies to BACKUP and RESTORE flavors that implicitly target an
offline importing/restoring table. A future PR should allow a user to
explicitly target an existing offline table undergoing an IMPORT.

Release justification: fixes bug where offline tables were not properly handled
in restore.

msbutler commented:
RFAL. I limited the scope of this PR to only deal with handling offline tables in restore. Currently on the backup side, all tables that come back online will get fully re-backed up. In a future PR, I will skip this re-backup step one tables that did not undergo a non-mvcc op during the inc backup interval.

msbutler commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -657,6 +657,25 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
 }
 
+// eligibleInProgressImport returns true if the descriptor represents a table with an in progress
+// import that started in a cluster finalized to version 22.2.
```
nit: should say: "import that started ina cluster finalized to version 22.2 and is getting restored on finalized 22.2 cluster"

msbutler replied:
the restoring cluster version check is superfluous as the descriptor's start time will never be above 0 in a mixed version restore. Perhaps I should instead conduct an assertion failure after checking that the import start time is greater than 0.

adityamaru commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -660,6 +660,27 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
 	return ok
 }
 
+// eligibleInProgressImport returns true if the descriptor represents a table with an in progress
```
nit: can we maybe clarify that this method is used on "backed up" descriptors and not descriptors in the cluster. To that effect what do you think about renaming this to `backedUpDescriptorsWithInProgressImportInto`, a bit wordy but maybe its okay.

msbutler replied:
Done.

adityamaru commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -660,6 +660,27 @@ func shouldPreRestore(table *tabledesc.Mutable) bool {
+	if table.GetInProgressImportStartTime() == 0 {
+		return false, nil
+	}
+	if !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.Start22_2) {
```
As we discussed on our call, I think this version check is a little disconnected from what it is guarding against. If we don't want descriptors to have a GetInProgressImportStartTime set in a mixed version we should guard it where we set it (I think you already do), instead of where we read it. I would delete this but if you feel better with it here its okay too.

msbutler replied:
Done. Deleted.

adityamaru commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -95,7 +95,7 @@ var restoreStatsInsertBatchSize = 10
 	codec keys.SQLCodec, kr *KeyRewriter, key roachpb.Key,
 ) (roachpb.Key, error) {
-	newKey, rewritten, err := kr.RewriteKey(append([]byte(nil), key...))
+	newKey, rewritten, err := kr.RewriteKey(append([]byte(nil), key...), 0)
```
nit: /* wallTime */

adityamaru commented on pkg/ccl/backupccl/restore_data_processor.go:
```diff
@@ -464,7 +464,12 @@ func (rd *restoreDataProcessor) processRestoreSpanEntry(
+		key.Key, ok, err = kr.RewriteKey(key.Key, key.Timestamp.WallTime)
+
+		if errors.Is(err, ErrImportingKeyError) {
+			// Elide importing keys
```
Lets add more to this comment, the logic is subtle so itll be nice to call out what and why we are eliding keys with this error.

adityamaru commented on pkg/ccl/backupccl/key_rewriter.go:
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
+// checkAndRewriteTableKey rewrites the table IDs in the key. It assumes that
+// any tenant ID has been stripped from the key so it operates with the system
+// codec. It is the responsibility of the caller to either remap, or re-prepend
+// any required tenant prefix.
```
Can we add a few sentences about the new parameter we have added, and what its used for.

adityamaru replied:
Also maybe mention that a zero value skips the checks associated with the wallTime.

adityamaru commented on pkg/ccl/backupccl/key_rewriter.go:
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
+		return nil, false, errors.Errorf("missing descriptor for table %d", tableID)
+	}
+
+	// Elide Importing Keys
```
Lets add more to this comment too, in particular the timestamp comparison and why its what we want in a restore.

adityamaru commented on pkg/ccl/backupccl/key_rewriter.go:
```diff
@@ -258,23 +258,33 @@ func (kr *KeyRewriter) RewriteKey(key []byte) ([]byte, bool, error) {
-// remap, or re-prepend any required tenant prefix.
-func (kr *KeyRewriter) rewriteTableKey(key []byte) ([]byte, bool, error) {
+// ErrImportingKeyError indicates the current key is apart of an in-progress import
+var ErrImportingKeyError = errors.New("Importing Key")
```
nit: I think all errors should be lower case unless its a proper noun, maybe "skipping rewrite of an importing key"?

adityamaru reviewed (commented):
Flushing out one round of comments, I need to look more carefully at the tests but can we maybe write a unit test in `key_rewriter_test.go` that isolates the new KR logic? The restore/import tests are super valuable but a unit test will give us more focussed control on testing the timestamp comparisons.

msbutler commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -708,6 +729,17 @@ func createImportingDescriptors(
 
 	for _, desc := range sqlDescs {
+
+		// TODO (msbutler) remove the schema_change condition once the online schema changer
```
Done.

msbutler commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -708,6 +729,17 @@ func createImportingDescriptors(
+
+		// TODO (msbutler) remove the schema_change condition once the online schema changer
+		// doesn't rely on OFFLINE state
+		if desc.Offline() && desc.GetDeclarativeSchemaChangerState() == nil {
```
Done.

msbutler reviewed (commented):
thanks for first pass! I think i got all your comments and added a keyrewriter test as well.

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:
```diff
@@ -0,0 +1,469 @@
+# ImportStartTime is not bound to the table's descriptor, therefore during
+# RESTORE AOST in-progress IMPORT, these tables should get thrown out.
+
+# TODO (msbutler): Currently, we still incrementally back up data from these
```
what jobs are we referring to here?

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:
```diff
@@ -0,0 +1,469 @@
+----
+
+
+# Conduct another set of incremental backups to ensure that _offline_ tables do _not_ fully backup
```
can you clarify what you mean by "fully backup" here?

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:
```diff
@@ -0,0 +1,469 @@
+# This test ensures that database and cluster backups properly
```
meta note: we're starting 6 servers in this test which could make it quite a slow test, I don't know if theres anything we can do about it given the lack of a "wipe everything" tool for cluster restore but we should run this under stress a few times.

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:
```diff
@@ -0,0 +1,469 @@
+d foofoo table 3 incremental
+
+
+# A restore on mixed version cluster should always fail on a backup that contains import times
```
The backup to database_upgrade won't contain descriptors with import time set right even though its post finalization, because the import was run in a pre-finalized state?

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:
```diff
@@ -0,0 +1,469 @@
+
+
+# Now actually Restore the backups taken from a mixed version chain
+new-server name=s6 share-io-dir=s1 allow-implicit-access
```
Can you just upgrade the previous server and reuse it, we can save a server startup I think.

adityamaru commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-imports:
```diff
@@ -0,0 +1,469 @@
+# Ensure the RESTOREs omit the tables with in progress imports (foo and foofoo)
+# as their descriptor will not have the import start time.
+restore aost=m0
+RESTORE FROM LATEST IN 'nodelocal://0/cluster/' AS OF SYSTEM TIME m0;
```
can we check RESTORE FROM DATABASE AOST m0 as well after dropping d?

adityamaru reviewed (changes requested):
Nice test coverage, and thanks for the comments!

msbutler commented on pkg/ccl/backupccl/key_rewriter.go:
```diff
@@ -276,7 +288,12 @@ func (kr *KeyRewriter) checkAndRewriteTableKey(key []byte, wallTime int64) ([]by
+	// GetInProgressImportStartTime), then this function returns an error if this
+	// key is a part of the import -- i.e. the key's walltime is greater than the
+	// import start time. It is up to the caller to handle this error properly.
+	fmt.Printf("walltime: %v; importTime %v \n", wallTime, desc.GetInProgressImportStartTime())
```
Ah nice catch. Removed.

msbutler commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_processor.go:
```diff
@@ -619,7 +619,7 @@ func (sip *streamIngestionProcessor) bufferSST(sst *roachpb.RangeFeedSSTable) er
 
 func (sip *streamIngestionProcessor) rekey(key roachpb.Key) ([]byte, error) {
-	rekey, ok, err := sip.rekeyer.RewriteKey(key)
+	rekey, ok, err := sip.rekeyer.RewriteKey(key, 0 /*wallTime*/)
```
Currently, passing a 0 walltime signals to RewriteKey to _never_ filter on walltime. A SIP may encounter a descriptor with an in progress import, and I don't want the SIP do key filtering. Passing a 0 walltime signals this. We could easily rename the var `walltimeFilter` to make this slightly more readable.

Alternatively, the caller could always pass a non-zero walltime and _also_ pass a `filterbyWalltime` bool to `makeKeyRewriter` that binds to the `KeyRewriter` struct which signals whether walltimes should be used for filtering in `RewriteKey`.

Again, I prefer what I've initially drafted, as it requires managing fewer internal vars and has no perf impact on SIP. It's up to the caller to know whether they want to filter on import start time via this low level function thats optimized for performance.

adityamaru replied:
> I don't want the SIP do key filtering

Have we thought this through though, as to why we don't want this behaviour to be uniform with the SIP? I am not very  familiar with the implications of this so @stevendanna can probably help make a decision here.

> It's up to the caller to know whether they want to filter on import start time

I agree with this sentiment but as you know I don't think this code is a reflection of that. The filtering should not be in this method but given our offline discussions, I think a happy compromise would be to unconditionally pass in the ts of the key that we are passing in if the semantics are well understood for the SIP.

Renaming the variable to something other than `timestamp` makes it even more likely that we pass in a timestamp that is not the keys timestamp, which we should never iiuc?

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores:
```diff
@@ -0,0 +1,132 @@
+exec-sql
+CREATE DATABASE b;
+USE b;
+CREATE TABLE baz (i INT PRIMARY KEY, s STRING);
```
Done. Let me know if this aligns with what you had in mind!

msbutler reviewed (commented):
you're welcome for all the new test coverage :D

msbutler commented:
@adityamaru RFAL!

adityamaru reviewed (approved):
LGTM with nits. Thanks for working with me through this, I think the test coverage is extensive so we should get this in and let it bake. Can you ping SQL Schema once this diff is borsing so that they are well informed when they're fixing their bug?

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores:
```diff
@@ -125,8 +147,295 @@ b2
+job paused at pausepoint
+
+
+# ensure user cannot override offline schema created by table restore into data database d
```
Ah good catch-- though I'm not sure d in the backed up cluster from L139 actually has the schema. Anyway, I added the following to clarify things, and this test still holds.
```
USE d;
DROP SCHEMA IF EXISTS me CASCADE;
```


Note that during  the beginning of restore execution, the schema is published to an offline state. I'm not sure how to implement the UX you want. How is the user supposed to override an already published schema? Seems like an issue for another day.

msbutler commented on pkg/ccl/backupccl/key_rewriter_test.go:
```diff
@@ -147,6 +147,40 @@ func TestKeyRewriter(t *testing.T) {
+			systemschema.NamespaceTable.GetID(), desc.GetPrimaryIndexID())
+
+		// If the passed in walltime is at or above the ImportStartWalltime, an error should return
+		_, _, err = newKr.RewriteKey(key, 2)
```
Done.

msbutler commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -718,6 +718,10 @@ func createImportingDescriptors(
 
 	oldTableIDs := make([]descpb.ID, 0)
 
+	// offlineSchemas are used to ensure that restoring databases descriptors do
```
Done.

msbutler commented on pkg/ccl/backupccl/testdata/backup-restore/in-progress-restores:
```diff
@@ -53,43 +68,50 @@ SET CLUSTER SETTING jobs.debug.pausepoints = '';
+b2 me schema <nil> incremental
+b2 greeting type <nil> incremental
+b2 _greeting type <nil> incremental
 b2 baz table 3 incremental
```
Answered offline.

msbutler commented:
TFTRS!!

bors r=adityamaru
